feat: recently viewed projects

eikek opened the pull request:
Implements returning recently viewed entities.

/deploy #persist

eikek committed:
Trial

eikek committed:
EndpointDocs stub for recently viewed entities

eikek committed:
Starting out slowly

eikek committed:
Refactoring: pass config as argument vs loading it everywhere

It also makes it possible to use it from tests with a different db
connection.

eikek committed:
Starting with querying

eikek committed:
file headers

eikek committed:
Add recently viewed projects query

eikek committed:
More tests

eikek committed:
Starting with dataset query

eikek committed:
Dataset Query

eikek committed:
Test updates

eikek committed:
Fix dataset query

eikek committed:
Add route to microservice

eikek committed:
Remove obsolete file

eikek committed:
Fix endpoint docs runtime errors

eikek committed:
One more test

eikek committed:
Truncate view dates to seconds

reduces likelyhood that two writes happen only few millis apart which
will currently yield in duplicate view dates

eikek committed:
Fix endpoint doc errors

jachro commented on entities-viewings-collector/src/main/scala/io/renku/entities/viewings/search/DatasetQuery.scala:
```diff
@@ -0,0 +1,129 @@
+               |        (GROUP_CONCAT(DISTINCT ?keyword; separator=',') AS ${v.keywords})
+               |        (GROUP_CONCAT(DISTINCT ?encodedImageUrl; separator=',') AS ${v.images})
+               |      WHERE {
+               |        Graph schema:Dataset {
```
Should we maybe use the `GraphClass.Datasets.id` as graph id here (and in few other places in the query)?

eikek replied:
Yes ofc! Forgot these occurrences to replace. Thanks

jachro commented on entities-viewings-collector/src/test/scala/io/renku/entities/viewings/search/DatasetQuerySpec.scala:
```diff
@@ -0,0 +1,137 @@
+
+  it should "return multiple datasets" in {
+    val project1 = renkuProjectEntities(visibilityPublic)
+      .withActivities(activityEntities(stepPlanEntities()))
```
I'm wondering if we really need to generate activities in these specs? If not, it's always fewer lines of code plus a tiny perf improvement during execution.

eikek replied:
right, probably not needed! I'll remove them

jachro commented on entities-viewings-collector/src/test/scala/io/renku/entities/viewings/search/DatasetQuerySpec.scala:
```diff
@@ -0,0 +1,137 @@
+
+    val dataset1 = project1.to[entities.RenkuProject.WithoutParent].datasets.head
+    val dataset2 = project2.to[entities.RenkuProject.WithoutParent].datasets.head
+    val userId1  = person1.maybeGitLabId.get
```
It's not a big deal but if we'd mix in `OptionValues` and use `.value` instead of `.get` we can eventually get slightly more informative failures when they happen.

jachro commented on entities-viewings-collector/src/test/scala/io/renku/entities/viewings/search/ProjectQuerySpec.scala:
```diff
@@ -0,0 +1,128 @@
+
+  it should "return one project entry if viewed multiple times" in {
+    val project = renkuProjectEntities(visibilityPublic)
+      .withActivities(activityEntities(stepPlanEntities()))
```
Personally, for the sake of simplicity, I'd maybe not add activities or datasets to these specs. Forget the comment if I missed something and they are needed here.

eikek replied:
I think I can remove the datasets here instead. I thought first, why not add them, just to also see if they are not returned. But can also remove

jachro commented on entities-viewings-collector/src/test/scala/io/renku/entities/viewings/search/SearchTestBase.scala:
```diff
@@ -0,0 +1,102 @@
+
+  lazy val tsClient = TSClient[IO](projectsDSConnectionInfo)
+
+  def storeProjectViewed(userId: GitLabId, dateViewed: Instant, path: projects.Path): Unit =
```
It's only in the latest PR for the viewing dates deduplication where I'm introducing an `EntityViewings` trait for the purpose of provisioning Viewing Events in our specs. We can maybe consider using them here one the other PR is merged.

jachro commented on entities-viewings-collector/src/test/scala/io/renku/entities/viewings/search/SearchTestBase.scala:
```diff
@@ -0,0 +1,102 @@
+  def printSparqlQuery(q: SparqlQuery) =
+    println(q.toString.split('\n').zipWithIndex.map(t => f"${t._2}%2d ${t._1}").mkString("\n"))
+
+  def projectDecoder: Decoder[List[SearchEntity.Project]] =
```
What do you think about using the `ResultsDecoder` in these decoders? We'd have the `results.bindings` path hardcoded in one place, then.

eikek replied:
Ah yes, thanks. I missed that.

jachro commented on knowledge-graph/src/main/scala/io/renku/knowledgegraph/MicroserviceRoutes.scala:
```diff
@@ -87,9 +90,23 @@ private class MicroserviceRoutes[F[_]: Async](
+    `GET /datasets/*` <+> `GET /entities/*` <+> `GET /projects/*` <+> `GET /users/*` <+> getRecentEntities
   }
 
+  private lazy val getRecentEntities: AuthedRoutes[Option[AuthUser], F] =
```
I guess we'd need to use the new `MaybeAuthUser` here but the compiler will complain anyway. Actually, is it true that a token is mandatory here? I believe we wouldn't be able to tell much about a current user if the token is not given.

eikek replied:
From what I understood here is that the user is required, right. So it probably needs to be `AuthUser` then, thanks!

jachro commented on knowledge-graph/src/main/scala/io/renku/knowledgegraph/entities/QueryParamDecoders.scala:
```diff
@@ -45,6 +48,37 @@ object QueryParamDecoders {
     val parameterName: String = "type"
   }
 
+  private implicit val recentlyViewedEntityTypesParamDecoder: QueryParamDecoder[RecentEntitiesFinder.EntityType] =
```
Do you think it's a good place for the new endpoint params to sit here together with the `entities` endpoint encoders?

eikek replied:
Well, don't have a strong opinion here. I thought that these two could probably share a lot of things, like page/limit and later also entityTypes etc. And they are _used_ outside of that package anyways. But can move them to the sub-package?

jachro replied:
Yes, you are right they share a lot. I was just looking at that from the perspective of having things tidy. But, yes, maybe a subpackage that is used by both endpoints would be a good idea.

eikek replied:
I now moved the class one level up. I noticed that it isn't used in that subtree anywhere but solely in the `MicroserviceRoutes` class. Hope that is ok, but can be changed any time obviously.

jachro commented on triples-generator/src/main/scala/io/renku/triplesgenerator/Microservice.scala:
```diff
@@ -64,6 +64,7 @@ object Microservice extends IOMicroservice {
     implicit0(acf: AccessTokenFinder[IO])        <- AccessTokenFinder[IO]()
     implicit0(rp: ReProvisioningStatus[IO])      <- ReProvisioningStatus[IO]()
     config                                       <- parseConfigArgs(args)
+    projectConnConfig                            <- ProjectsConnectionConfig[IO](config)
```
I'm wondering if it wouldn't be possible/good to pass the `ProjectsConnectionConfig` in a bit more implicit way. I somehow liked the pureness of these definitions here. The other argument for doing so is that it's rather common for all the instantiations here that they don't need the config for themselves but rather pass them down. What do you think Eike?

eikek replied:
We can discuss later today if you want. I found it not so nice that every little class is re-reading the config from the filesystem again and again; I tink it terms of "pureness" it is better to pass it along than reading it every time. But the main motivation was to allow different configs for testing purposes. Personally I'd prefer reading the config once and then pass it to where it is needed (or parts of it ofc). Also personal preference :) I don't like using implicits for that, I think it gets easily overused and then very hard to read. If some parts only pass it down without needing actually the data, then perhaps we could introduce the service in need as dependency and not the config for creating it?

eikek committed:
Address review comments

eikek committed:
Fix compile errors due to changes on base branch

eikek committed:
Move query parameter decoders up one level

They are not used by the endpoint implementation, but only in the
class that defines the route endpoints for http4s.
